src/RSHiScores.php

```diff
@@ -241,11 +230,105 @@ private static function lookup( $api, $player, $skill, $type ) {
+			return json_encode( $data );
 		} else {
-			return self::parse( $data, $skill, $type );
+			return self::getSingleSkillData( $data, $skill, $type, $skillIsInt );
+		}
+	}
+
+	/**
+	 * Do some post-processing of the data from the endpoint.
+	 * Most notably html-escape data from untrusted API
+	 *
+	 * @param string $data The received data
+	 * @return array The processed data
+	 */
+	private static function postFetch( $data ) {
+		$data = json_decode( $data, true );
+		if ( !is_array( $data ) ) {
+			// Error: Endpoint returned invalid json
+			throw new Exception( wfMessage( 'rshiscores-error-invalid-json' ) );
+		}
+
+		// Index all skills/activities in flat array by lowercase key, for easy lookup.
+		$parsedData = [];
+		foreach ( $data as $skillOrActivity => $stats ) {
+			foreach ( $stats as $stat ) {
+				if ( isset( $stat['name'] ) ) {
+					$parsedData[ strtolower( $stat[ 'name' ] ) ] = $stat;
+				}
+			}
+		}
+
+		return self::escapeStrings( $parsedData );
+	}
+
+	/**
+	 * General-purpose html-escaper.
+	 * Recurses into an array and escapes all keys and scalar values encountered.
+	 *
+	 * @param array|string $arrayOrString The string to escape.
+	 * @return array|string The escaped result.
+	 */
+	private static function escapeStrings( $arrayOrString ) {
+		if ( is_scalar( $arrayOrString ) ) {
+			return htmlspecialchars( $arrayOrString, ENT_QUOTES );
+		}
+		if ( is_array( $arrayOrString ) ) {
+			$rtr = [];
+			foreach ( $arrayOrString as $key => $value ) {
+				$rtr[ self::escapeStrings( $key ) ] = self::escapeStrings( $value );
+			}
+			return $rtr;
+		}
+		// This should not happen
+		throw new Exception( 'rshiscores-error-unexpected-json' );
+	}
+
+	/**
+	 * Get data for specific skill and type from $data.
+	 *
+	 * @param array $data The data fetched from the endpoint, and processed by self::postFetch
+	 * @param string $skill The skill to search for in the data.
+	 * @param string $type The type (xp/rank/score/level/self::DEFAULT_TYPE) of data to get for the skill.
+	 * @param bool $fromInt Should $skill be interpreted as an integer?
+	 *
+	 * @return string The requested data
+	 *
+	 * @throws Exception If $skill or $type could not be found, or if endpoint returned unexpected results
+	 */
+	private static function getSingleSkillData( $data, $skill, $type, $fromInt ) {
+		if ( $fromInt ) {
+			if ( (int) $skill >= count($data) ) {
+				throw new Exception( wfMessage( 'rshiscores-error-unknown-skill' ) );
+			}
+			// Semi-backwards compatibility: If skill is integer "n", take the "n"th element of the data.
+			$skill = array_keys( $data )[ $skill ];
+		} else {
+			// Case-insensitive, use same processing as self::postFetch did
+			$skill = self::escapeStrings( strtolower( $skill ) );
+		}
+		if ( !isset( $data[ $skill ] ) ) {
+			// Error: Skill/activity is unknown. Maybe they changed the spelling?
+			throw new Exception( wfMessage( 'rshiscores-error-unknown-skill' ) );
+		}
+		if ( $type === self::DEFAULT_TYPE ) {
```

Comment:
I could also do `$type = array_keys( $data[ $skill ] )[1];` instead, maybe? Not sure...